### PR TITLE
#1001 auto mapping support for update methods

### PR DESCRIPTION
**Pull request by sjaakd**
just a testcase to start with

### CODE REVIEW

**Comment by gunnarmorling** [comment]
ready for review.

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 1]
```diff
@@ -638,16 +639,33 @@ private Assignment forgeMapping(SourceRHS sourceRHS) {
...
-                targetType,
-                method.getMapperConfiguration(),
-                method.getExecutable(),
-                method.getContextParameters(),
-                getForgedMethodHistory( sourceRHS )
-            );
+            List<Parameter> parameters = new ArrayList( method.getContextParameters() );
+            ForgedMethod forgedMethod;
+            if ( method.isUpdateMethod() ) {
+                parameters.add( Parameter.forForgedMappingTarget( targetType ) );
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: I would just add the new parameter if the `method` is update method. Everything else can stay as it was before. 

If the `mappingTargetParameter` is not `null` and it is when we do an update method, then we return it's type. The `TypeKind.VOID` is obsolete

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/ForgedMethod.java [hunk 2]
```diff
@@ -303,4 +306,14 @@ public int hashCode() {
...
         return result;
     }
+
+    private Parameter determineMappingTargetParameter(Iterable<Parameter> parameters) {
+        for ( Parameter parameter : parameters ) {
+            if ( parameter.isMappingTarget() ) {
+                return parameter;
+            }
+        }
+
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/ForgedMethod.java]: 2 empty lines

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/ForgedMethod.java [hunk 3]
```diff
@@ -303,4 +306,14 @@ public int hashCode() {
         result = 31 * result + ( name != null ? name.hashCode() : 0 );
         return result;
     }
+
+    private Parameter determineMappingTargetParameter(Iterable<Parameter> parameters) {
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/ForgedMethod.java]: Just a not for the future, we can extract this construct to some util. We use it in multiple places 😄

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java (hunk 1 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: I think its more cleaner to return void. I struggled in the past with update methods that have both a return type and a mapping target. I still don't like that :).

Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: I agree with you. I forgot the `getReturnType`.

However, I will still reduce the amount of code. By either new variable which will be either the `targetType` or `TypeKind.VOID` or with a ternary on the same line 😄

Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: Ok.. will update.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/ForgedMethod.java (hunk 3 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/ForgedMethod.java]: I could do it as static on Parameter. Just like @agudian did for context parameters as well.

Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/ForgedMethod.java]: As you wish. We can also do it as part of another issue

**Comment by sjaakd** [comment]
@filiphr : I covered your remarks and did some small additional refactoring. Have a look if you agree. After I'll continue to merge.

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/Parameter.java [hunk 4]
```diff
@@ -125,6 +125,15 @@ public static Parameter forElementAndType(VariableElement element, Type paramete
             ContextPrism.getInstanceOn( element ) != null );
     }
 
+    public static Parameter forForgedMappingTarget( Type parameterType) {
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/Parameter.java]: 👍🙂

Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/Parameter.java]: Just the formatting looks a bit off. Is there a space after `(`?

**Comment by filiphr** [comment]
@sjaakd everything looks fine, there is only one small formatting issue (maybe it is not and I am not seeing it correctly 😄)

**Comment by sjaakd** [comment]
@agudian, @filiphr : thanks for reviewing

**Comment by sjaakd** [comment]
@filiphr : fixed the spacing

**Pull request merged**